File: nvxapp.client/src/nvx/ClientServer-Service/GestionePresenze/Dip_GG_Richiesta/Models/dip-gg-richiesta-model.ts
import { ModelResult } from "../../../ModelsBase/model-result";
import { Dip_GG_GiustificativiModel } from "../../Dip_GG_Giustificativi/Models/dip-gg-giustificativi-model";
import { Dip_GG_TimbraturaModel } from "../../Dip_GG_Timbratura/Models/dip-gg-timbratura-model";

/************************************************************/
/******************** MODEL DATA ****************************/
/************************************************************/

export class Dip_GG_RichiestaModel {

  id!: number;
  idDip_RapportoLavoro!: number;

  data!: string;
  dataA!: string;

  richiestaTipo!: TipoRichiesta;
  
  // Campo per oggetto JSON
  dati?: string;  // contiene oggetti di tipo  Dip_GG_Richiesta_Body_Timbratura,Dip_GG_Richiesta_Body_Giustificativo, Dip_GG_Richiesta_Body_NotaSpesa

  richiestaStato!: StatoRichiesta;
  richiestaApprovazioneData: Dip_GG_Richiesta_Stato_Cronology[];

  revocaStato?: StatoRichiesta;
  revocaApprovazioneData: Dip_GG_Richiesta_Stato_Cronology[];

}
export class Dip_GG_Richiesta_Stato_Cronology {
  idAspNetUsers!: string;
  richiestaStato!: StatoRichiesta;
  data!: Date;
}
export class Dip_GG_Richiesta_Body_Timbratura {
  hhmm!: string;
}
export class Dip_GG_Richiesta_Body_Giustificativo {
  hhmm!: string;
  allDay!: boolean;
  idPar_Giustificativi!: number;
}
export class Dip_GG_Richiesta_Body_NotaSpesa {
}


export enum TipoRichiesta {
  Timbratura,
  Giustificativo,
  NotaSpesa,
  ApprovazioneStraordinario
}
export enum StatoRichiesta {
  Diretta,

  Immessa,
  Cancellata,
  Rifiutata,
  ApprovazioneInCorso,
  ParzialmenteApprovata,

  Approvata
}


/************************************************************/
/******************** DATI INPUT X API **********************/
/************************************************************/

export class Dip_GG_Richiesta_GetAll4User_InModel {
  public idAspNetUsers?: string
  public year: number;
  public month: number;

  constructor() {}
}
export class Dip_GG_Richiesta_GetAll4User_OutModel extends ModelResult {

  public dip_GG_Richiesta: Dip_GG_RichiestaModel[];

}


export class Dip_GG_Richiesta_GetAll4Admin_InModel {
  public year: number;
  public month: number;

  constructor() { }
}
export class Dip_GG_Richiesta_GetAll4Admin_OutModel extends ModelResult {

  public dip_GG_Richiesta: Dip_GG_RichiestaModel[];

}


export class Dip_GG_Richiesta_Send_InModel {
  public excludeRicalc: boolean;
  public idAspNetUsers?: string
  public dip_GG_Richiesta: Dip_GG_RichiestaModel;
  public fromHR: boolean;
  constructor() {
    this.dip_GG_Richiesta = new Dip_GG_RichiestaModel(); 
  }
}
export class Dip_GG_Richiesta_Send_OutModel extends ModelResult {

  public dip_GG_Richiesta: Dip_GG_RichiestaModel[];
  public dip_GG_Timbratura: Dip_GG_TimbraturaModel[];
  public dip_GG_Giustificativi: Dip_GG_GiustificativiModel[];

  
  constructor() {
    super();
  }

}


export class Dip_GG_Richiesta_SetState_InModel {

  public excludeRicalc: boolean;
  public richiestaStato!: StatoRichiesta;
  public idDip_GG_Richiesta: number[] = [];
  public fromHR: boolean;

  constructor() {
    
  }
}
export class Dip_GG_Richiesta_SetState_OutModel extends ModelResult {

  public dip_GG_Richiesta: Dip_GG_RichiestaModel[] = [];
  public dip_GG_Timbratura: Dip_GG_TimbraturaModel[] = [];
  public dip_GG_Giustificativi: Dip_GG_GiustificativiModel[] = [];


  constructor() {
    super();
    //this.dip_GG_Richiesta = new Dip_GG_RichiestaModel();
  }

}


export class Dip_GG_Richiesta_Get_4Calculation_InModel {
  public usersId: string[] = [];
  public dal!: string;
  public al!: string;
}
export class Dip_GG_Richiesta_Get_4Calculation_OutModel extends ModelResult {
  public dip_GG_Richiesta: Dip_GG_RichiestaModel[] = [];
}

File: nvxapp.server.service/ClientServer-Service/GestionePresenze/Dip_GG_RichiestaService/Models/Dip_GG_RichiestaModel.cs
using nvxapp.server.data.Entities.Tenant;
using nvxapp.server.service.ClientServer_Service.GestionePresenze.Dip_GG_GiustificativiService.Models;
using nvxapp.server.service.ClientServer_Service.GestionePresenze.Dip_GG_TimbraturaService.Models;
using nvxapp.server.service.ClientServer_Service.ModelsBase;

namespace nvxapp.server.service.ClientServer_Service.GestionePresenze.Dip_GG_RichiestaService.Models
{




    public class Dip_GG_RichiestaModel : HashModel
    {
        public required int Id { get; set; }
        public required int IdDip_RapportoLavoro { get; set; }

        [HashField]
        public string Data { get; set; } = string.Empty;
        [HashField]
        public string DataA { get; set; } = string.Empty;

        [HashField]
        public TipoRichiesta RichiestaTipo { get; set; }
        // Campo per oggetto JSON
        [HashField]
        public required string Dati { get; set; } = string.Empty;

        [HashField]
        public StatoRichiesta RichiestaStato { get; set; }
        [HashField]
        public List<Dip_GG_Richiesta_Stato_Cronology>? RichiestaApprovazioneData { get; set; } = new List<Dip_GG_Richiesta_Stato_Cronology>();
        [HashField]
        public StatoRichiesta? RevocaStato { get; set; }
        [HashField]
        public List<Dip_GG_Richiesta_Stato_Cronology>? RevocaApprovazioneData { get; set; } = new List<Dip_GG_Richiesta_Stato_Cronology>();

        // Campo per oggetto JSON
        [HashField]
        public string? CronologiaApprovazione { get; set; }

    }

    public class Dip_GG_Richiesta_Body_Timbratura
    {
        public required string hhmm { get; set; } = string.Empty;
    }
    public class Dip_GG_Richiesta_Body_Giustificativo
    {
        public required string hhmm { get; set; } = string.Empty;
        public required bool AllDay { get; set; } = false;
        public required int IdPar_Giustificativi { get; set; } = 0;
    }
    public class Dip_GG_Richiesta_Body_NotaSpesa
    {

    }


    public class Dip_GG_Richiesta_GetAll4User_InModel
    {
        public string? IdAspNetUsers { get; set; }
        public int Year { get; set; } = 0;
        public int Month { get; set; } = 0;
    }
    public class Dip_GG_Richiesta_GetAll4User_OutModel : ModelResult
    {
        public List<Dip_GG_RichiestaModel> Dip_GG_Richiesta { get; set; } = new List<Dip_GG_RichiestaModel>();
    }


    public class Dip_GG_Richiesta_GetAll4Admin_InModel
    {
        public int Year { get; set; } = 0;
        public int Month { get; set; } = 0;
    }
    public class Dip_GG_Richiesta_GetAll4Admin_OutModel : ModelResult
    {
        public List<Dip_GG_RichiestaModel> Dip_GG_Richiesta { get; set; } = new List<Dip_GG_RichiestaModel>();
    }




    public class Dip_GG_Richiesta_Send_InModel
    {
        public string? IdAspNetUsers { get; set; }
        public Dip_GG_RichiestaModel? Dip_GG_Richiesta { get; set; }
        public Boolean FromHR { get; set; }
    }
    public class Dip_GG_Richiesta_Send_OutModel : ModelResult
    {
        public Dip_GG_Richiesta_Send_OutModel()
        {
            Dip_GG_Richiesta = new List<Dip_GG_RichiestaModel>();
            Dip_GG_Timbratura = new List<Dip_GG_TimbraturaModel>();
            Dip_GG_Giustificativi = new List<Dip_GG_GiustificativiModel>();
        }

        public List<Dip_GG_RichiestaModel> Dip_GG_Richiesta { get; set; } = new List<Dip_GG_RichiestaModel>();
        public List<Dip_GG_TimbraturaModel> Dip_GG_Timbratura { get; set; } = new List<Dip_GG_TimbraturaModel>();
        public List<Dip_GG_GiustificativiModel> Dip_GG_Giustificativi { get; set; } = new List<Dip_GG_GiustificativiModel>();
    }


    public class Dip_GG_Richiesta_SetState_InModel
    {
        public List<int> IdDip_GG_Richiesta { get; set; } = new List<int>();

        public StatoRichiesta RichiestaStato { get; set; }
        public Boolean FromHR { get; set; }

    }
    public class Dip_GG_Richiesta_SetState_OutModel : ModelResult
    {
        // dati modificati
        public List<Dip_GG_RichiestaModel> Dip_GG_Richiesta { get; set; } = new List<Dip_GG_RichiestaModel>();
        public List<Dip_GG_TimbraturaModel> Dip_GG_Timbratura { get; set; } = new List<Dip_GG_TimbraturaModel>();
        public List<Dip_GG_GiustificativiModel> Dip_GG_Giustificativi { get; set; } = new List<Dip_GG_GiustificativiModel>();

    }

    public class Dip_GG_Richiesta_Get_4Calculation_InModel
    {
        public List<string> UsersId { get; set; } = new List<string>();
        public DateTime Dal { get; set; }
        public DateTime Al { get; set; }
    }

    public class Dip_GG_Richiesta_Get_4Calculation_OutModel : ModelResult
    {
        public List<Dip_GG_RichiestaModel> Dip_GG_Richiesta { get; set; } = new List<Dip_GG_RichiestaModel>();
    }

}
